036-kraft-mode.md

```diff
@@ -0,0 +1,103 @@
+# KRaft support: ZooKeeper-less Kafka
+
+## Motivation
+
+The Apache Kafka project is working on removing its dependency on ZooKeeper.
+Because of the complexity of this task, the work is in progress already for more than one year.
+It was stared with the [Kafka Improvement Proposal 500 (KIP-500)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum) and will continue for at least six to twelve months until all work is done and it is production-ready.
+Step by step, it removes the different dependencies on ZooKeeper.
+From the smaller things such as CLI tools or management APIs up to the complicated tasks such as quorum management and metadata management.
+It is based on _Apache Kafka Raft_ consensus protocol, which is also called _KRaft_.
+This is what gives the name to this proposal as well as to the feature gate discussed later.
+
+Strimzi aims to support the ZooKeeper-less Kafka and the KRaft protocol.
+Since the announcement of the KIP-500, we are making sure that Strimzi is as ready as possible for when the KRaft mode becomes production-ready.
+This includes for example adopting the new APIs based on Kafka instead of ZooKeeper in components such as User Operator or Cruise Control.
+But right now, you cannot really use Strimzi to run Kafka with the KRaft mode enabled.
+That makes it hard to continue working on some of the additional tasks needed to have Strimzi production-ready with KRaft in the future.
+For example the work on the readiness / liveness probes, Topic Operator improvements, upgrades etc.
+All of these tasks will be easier to work on with Strimzi supporting the KRaft mode out-of-the-box.
+
+## Proposal
+
+This proposal suggests to add provisional support for the KRaft mode.
+It will be disabled by default and will be protected by a new feature gate.
+When the feature gate is enabled, it will use the KRaft mode to deploy the Kafka cluster.
+It will simplify development and testing of new Strimzi features in the KRaft mode.
+And it would also allow to help with testing of the KRaft mode it self and raise any issues in the Apache Kafka project.
+
+The KRaft support proposed here is called provisional, because the API in the Kafka CR for configuring the Kafka nodes is not final and will be changed later.
+Most of the code implemented by this proposal is expected to be used for the final production-ready implementation.
+
+**The KRaft mode support suggested by this proposal will not be production-ready and should not be used outside of development.**
+
+### API changes
+
+In the initial implementation, there will be no changes to the `Kafka` custom resource and its API.
+When the KRaft mode is enabled, the operator will ignore any of the unsupported but required fields such as `.spec.zookeeper`.
+It will also validate the resource for any other unsupported features (such as authorization, JBOD storage etc.).
```

Comment:
I was not completely sure what is the best for this one. My branch currently makes the `.spec.zookeeper` optional. That is of course not great for the regular users who will not get the proper validation on the Kubernetes side.

So this proposal suggests to keep it as required, but ignore it. But just to make it clear, since it has some required fields inside `.spec.zookeeper`, you cannot do just `zookeeper: {}`. You need to configure the storage as well for example. So it is a bit weird.

Please let me know what you think about this. If needed, I can make it optional for everyone as it is in my branch and validate it only inside the operator code.

Reply:
I agree that it makes sense to leave `spec.zookeeper` as required for now.

Reply:
Yeah I guess that a warning in the log which says that `spec.zookeeper` is ignored would be enough.

Reply:
I think keeping this helps to make it *very clear* that this is not a production feature.